pipeline-ingress service

jgrillo-grapl opened the pull request:
<!-- Thank you for your contribution to Grapl! Please take some time
to fill out this short template to help us review and understand your
PR. -->

### Which issue does this PR correspond to?

https://github.com/grapl-security/issue-tracker/issues/580

### What changes does this PR make to Grapl? Why?

This PR adds a pipeline-ingress service which accepts raw log messages over gRPC and publishes `Envelope<RawLog>` messages to the `raw-logs` kafka topic.

### How were these changes tested?

Automated tests

jgrillo-grapl committed:
first draft of a pipeline-ingress service.

I simplified the pipeline ingress API (does not actually need to be
bidirectionally streaming, can be synchronous) and implemented an encapsulated
protocol buffer client/server in rust-proto new which does not expose types
generated by the protocol buffer compiler.

I also added a pipeline-ingress app which serves the pipeline ingress server and
implements its business logic interface. A subsequent commit will exercise these
components in tests both in rust-proto-new (to test the gRPC machinery) and
pipeline-ingress (to test the business logic).

colin-grapl commented on src/rust/rust-proto-new/src/graplinc/grapl/api/pipeline_ingress/v1beta1.rs:
```diff
@@ -137,3 +137,199 @@ impl SerDe for PublishRawLogsResponse {
+                   PipelineIngressProto::new(self.api_server)
+               ))
+               .serve(self.addr)
+               .map_err(|e| ConfigurationError::from(e))
```
I think you can replace this with ?

jgrillo-grapl replied:
Yes this closure is redundant, will fix.

jgrillo-grapl committed:
use the singular form to make the intent clear

jgrillo-grapl commented on src/rust/rust-proto-new/src/graplinc/grapl/api/pipeline_ingress/v1beta1.rs:
```diff
@@ -137,3 +137,199 @@ impl SerDe for PublishRawLogsResponse {
+        E: ToString + 'static
+    {
+        api_server: T,
+        _e: PhantomData<E>
```
`E` is used in `T`. Unfortunately, if I don't include it here I get a compiler error complaining that `E` doesn't exist. I do _need_ `E` here because it's the error type passed in via the `PipelineIngressApi<E>` interface.

colin-grapl replied:
Oops, I missed that it was in T. Yep, that's correct.

jgrillo-grapl committed:
WIP: testing

build out pipeline ingress server, client, app, and add an integration test for the transport layer

jgrillo-grapl committed:
WIP testing

jgrillo-grapl committed:
first draft of healthchecks

jgrillo-grapl committed:
formatting and lints

jgrillo-grapl committed:
use bind(0) to assign random ephemeral port in tests

jgrillo-grapl committed:
use ipv4 addresses because docker >_<

jgrillo-grapl committed:
add telemetry and placeholder integration test

jgrillo-grapl committed:
wire up integration-tests-new (and add Cargo.lock because I forgot in previous commit)

jgrillo-grapl committed:
run integration-tests-new in the verify and testing pipelines

jgrillo-grapl committed:
WIP add log message to determine that the server is shutting down

jgrillo-grapl committed:
fix a bug which was preventing pipeline-ingress from starting and add a whole bunch of instrumentation

jgrillo-grapl committed:
make the integration tests run

jgrillo-grapl committed:
make format

jgrillo-grapl committed:
update CODEOWNERS

jgrillo-grapl committed:
write integration test to exercise pipeline-ingress

jgrillo-grapl committed:
fix lints

jgrillo-grapl commented on nomad/grapl-core.nomad:
```diff
@@ -81,6 +81,11 @@ variable "engagement_creator_queue" {
 
+variable "kafka_bootstrap_servers" {
+  type        = string
+  description = "The URL of the Kafka bootstrap servers."
```
It's a little complicated. In general, the underlying `bootstrap.servers` configuration parameter in `librdkafka` takes a comma-separated list of `host:port` pairs. However in our particular case, for now, we'll only ever have one `host:port` pair.

- In local testing environments we only have a single broker
- In AWS we use confluent cloud, and they put their clusters behind a load balancer (which presents as one endpoint). See https://stackoverflow.com/a/68196232.

I believe this string gets passed transparently to librdkafka, so there's no extra work required on our end to support multiple bootstrap servers. I'll update this docstring to reflect the librdkafka documentation (see the `bootstrap.servers` entry here: https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md)

jgrillo-grapl replied:
done

jgrillo-grapl commented on nomad/grapl-core.nomad:
```diff
@@ -152,6 +157,26 @@ variable "organization_management_db_password" {
   description = "What is the password for the organization management database?"
 }
 
+variable "pipeline_ingress_healthcheck_polling_interval_ms" {
```
It's nice in tests to be able to have subsecond precision here. I doubt we'll need subsecond precision in prod but it didn't seem like any harm to allow it.

jgrillo-grapl commented on src/rust/pipeline-ingress/src/main.rs:
```diff
@@ -0,0 +1,175 @@
+        // TODO: trace_id should be generated at the edge. This service is
+        // currently "the edge" but that won't be true forever. When there is an
+        // actual edge service, that service should be responsible for
+        // generating the trace_id.
```
yes

jgrillo-grapl commented on src/rust/pipeline-ingress/src/main.rs:
```diff
@@ -0,0 +1,175 @@
+    tracing::info!("kafka producer configured successfully");
+
+    tracing::info!("configuring gRPC server");
+    let (server, _shutdown_tx) = PipelineIngressServer::new(
```
`shutdown_tx` would be used once we implement a signal handler to ensure graceful shutdown. It is present but ignored here because when I wrote it `let (server, _) = ...` immediately dropping the `shutdown_tx` caused the server to shutdown immediately. So it must be present, but for now we don't use it.

jgrillo-grapl commented on src/rust/pipeline-ingress/tests/integration_test.rs:
```diff
@@ -0,0 +1,137 @@
+            .with(log_layer)
+            .with(tracing_opentelemetry::layer().with_tracer(tracer))
+            .init();
+
```
It should be at some point. The question is where to put it? Some type of overarching service abstraction would be the sensible place I think. I've been putting this sort of thing off until I have a concrete need (e.g. more than one service) to let the requirements drive the code design rather than my own "cleverness" or "creativity". I suspect it'll be something to do sooner rather than later though.

jgrillo-grapl commented on src/rust/pipeline-ingress/src/main.rs:
```diff
@@ -0,0 +1,175 @@
+    let tracer = opentelemetry_jaeger::new_pipeline()
+        .with_service_name("pipeline-ingress")
+        .install_batch(opentelemetry::runtime::Tokio)
+        .unwrap();
```
Thanks, I missed this when I copy-pasta'd from the `grapl-config` crate, which I deliberately do not depend on here because it would drag in a whole bunch of heavy dependencies I don't want or need.

jgrillo-grapl replied:
fixed

jgrillo-grapl commented on src/rust/pipeline-ingress/tests/integration_test.rs:
```diff
@@ -0,0 +1,137 @@
+        .await
+        .expect("received error response");
+
+    assert!(res.created_time < SystemTime::now());
```
Good point, I initially had this with the inequality pointing the other way to deliberately trigger a test failure for my own sanity checking purposes. I should have just deleted the assert as it's not an invariant we actually care about, and as you note will likely cause a headache in AWS.

jgrillo-grapl replied:
fixed

jgrillo-grapl commented on src/rust/pipeline-ingress/tests/integration_test.rs:
```diff
@@ -0,0 +1,137 @@
+            .expect("could not subscribe to the raw-logs topic")
+            .any(|res: Result<Envelope<RawLog>, ConsumerError>| async move {
+                let metadata = res.expect("error consuming message from kafka").metadata;
+                metadata.tenant_id == tenant_id && metadata.event_source_id == event_source_id
```
Can't hurt

jgrillo-grapl replied:
done

jgrillo-grapl commented on src/rust/rust-proto-new/src/graplinc/grapl/api/pipeline_ingress/v1beta1.rs:
```diff
@@ -133,7 +143,515 @@ impl SerDe for PublishRawLogsResponse {
+    pub struct PipelineIngressServer<T, E, H, F>
+    where
+        T: PipelineIngressApi<E> + Send + Sync + 'static,
+        E: ToString + Send + Sync + 'static,
```
Yes `E` is the type of the error returned from the user-supplied API implementation.

jgrillo-grapl commented on src/rust/rust-proto-new/src/graplinc/grapl/api/pipeline_ingress/v1beta1.rs:
```diff
@@ -133,7 +143,515 @@ impl SerDe for PublishRawLogsResponse {
+        T: PipelineIngressApi<E> + Send + Sync + 'static,
+        E: ToString + Send + Sync + 'static,
+        H: Fn() -> F + Send + Sync + 'static,
+        F: Future<Output = Result<HealthcheckStatus, HealthcheckError>> + Send,
```
`F` only appears in the line above which defines the type `H` of the healthcheck function. I tried something like this but couldn't make it compile:

```rust
    H: Fn() -> impl Future<...> + ...
```

jgrillo-grapl commented on src/rust/pipeline-ingress/src/main.rs:
```diff
@@ -0,0 +1,175 @@
+
+#[tracing::instrument(err)]
+async fn handler() -> Result<(), ConfigurationError> {
+    let socket_address = std::env::var("PIPELINE_INGRESS_BIND_ADDRESS")?;
```
Yes, in that case we return a `std::io::Error` on L128 which gets converted to a `ConfigurationError::SocketError`.

jgrillo-grapl committed:
add newline to separate the two jobs

jgrillo-grapl committed:
move environment variable stuff out of the libraries and into the application layer, simplify some things, and address review comments

jgrillo-grapl committed:
improve bootstrap servers docstring and fix env vars

jgrillo-grapl commented on src/rust/rust-proto-new/tests/integration_test.rs:
```diff
@@ -235,18 +259,174 @@ proptest! {
+//       application layer.
+//   (2) To send a poison pill to our mocked application layer, to
+//       trigger an error response.
+const TENANT_ID: &'static str = "f000b11e-b421-4ffe-87c2-a963b77fd8e9";
```
I think these hardcoded IDs should be sufficiently random to avoid collisions even in the case where they're subject to things like shrinking in `proptest` tests. For example, if I made this ID all `0`s or all `f`s I might expect a proptest or similar strategy to hit those values while shrinking.

While this collision wouldn't be catastrophic for the `TENANT_ID`, it would cause a false-positive with the `BAD_EVENT_SOURCE_ID`.

jgrillo-grapl commented on src/rust/rust-proto-new/src/graplinc/grapl/api/pipeline_ingress/v1beta1.rs:
```diff
@@ -133,7 +143,515 @@ impl SerDe for PublishRawLogsResponse {
+            let inner_request: PublishRawLogRequest = match request.into_inner().try_into() {
+                Ok(request) => request,
+                Err(e) => return Err(Status::unknown(e.to_string())),
+            };
```
fixed

jgrillo-grapl commented on src/rust/rust-proto-new/tests/integration_test.rs:
```diff
@@ -235,18 +259,174 @@ proptest! {
+        let tenant_id = Uuid::parse_str(TENANT_ID).expect("failed to parse tenant_id");
+        assert!(request.tenant_id == tenant_id);
+
+        if request.event_source_id == tenant_id {
```
done

jgrillo-grapl commented on src/rust/rust-proto-new/tests/integration_test.rs:
```diff
@@ -235,18 +259,174 @@ proptest! {
+    if let Ok(res) = ctx
+        .client
+        .publish_raw_log(PublishRawLogRequest {
+            event_source_id: tenant_id,
```
fixed

jgrillo-grapl commented on src/rust/rust-proto-new/src/graplinc/grapl/api/pipeline_ingress/v1beta1.rs:
```diff
@@ -133,7 +143,515 @@ impl SerDe for PublishRawLogsResponse {
+            let response: PublishRawLogResponseProto = match response.try_into() {
+                Ok(res) => res,
+                Err(e) => return Err(Status::unknown(e.to_string())),
+            };
```
fixed

jgrillo-grapl commented on src/rust/rust-proto-new/src/graplinc/grapl/api/pipeline_ingress/v1beta1.rs:
```diff
@@ -133,7 +143,515 @@ impl SerDe for PublishRawLogsResponse {
+                Err(e) => return Err(Status::unknown(e.to_string())),
+            };
+
+            Ok(Response::new(response))
```
done

jgrillo-grapl commented on src/rust/kafka/src/lib.rs:
```diff
@@ -126,10 +126,14 @@ impl<T: SerDe> Producer<T> {
+                        tracing::trace!(
+                            "wrote message to partition {} offset {}",
+                            partition,
+                            offset
```
fixed

jgrillo-grapl commented on .buildkite/pipeline.testing.yml:
```diff
@@ -22,3 +22,18 @@ steps:
       - .buildkite/scripts/grapl_testing_stack/run_parameterized_job.sh e2e-tests 6
     artifact_paths:
       - "test_artifacts/**/*"
+  - label: ":aws: Integration tests in AWS"
```
done

jgrillo-grapl committed:
make the integration tests pass

jgrillo-grapl committed:
fmt

jgrillo-grapl committed:
remove unused deps

colin-grapl commented on src/rust/pipeline-ingress/src/main.rs:
```diff
@@ -0,0 +1,209 @@
+async fn handler() -> Result<(), ConfigurationError> {
+    let socket_address = std::env::var("PIPELINE_INGRESS_BIND_ADDRESS")?;
+
+    let healthcheck_polling_interval_ms =
+        std::env::var("PIPELINE_INGRESS_HEALTHCHECK_POLLING_INTERVAL_MS")?.parse()?;
+
+    let bootstrap_servers = std::env::var("KAFKA_BOOTSTRAP_SERVERS")?;
+    let sasl_username = std::env::var("KAFKA_SASL_USERNAME")?;
+    let sasl_password = std::env::var("KAFKA_SASL_PASSWORD")?;
```
These seem like they might be better off in a structopt configuration struct?

jgrillo-grapl replied:
I think we can (and probably _should_) punt on this type of thing until we figure out a good service abstraction. In particular, it appears `structopt` is deprecated. Their documentation suggests that we use `clap` instead. Additionally, we'll need to do a bunch of work to consolidate (and tease apart, where needed) the protobuf and kafka abstractions to the point that they can work with some kind of a "grapl service builder" type thing. Moreover, we need an idea of service lifecycle a la https://github.com/google/guava/wiki/ServiceExplained, signal handling, logging/tracing configuration, etc.

There are some clear answers here (like we should definitely use tower because it interoperates with tonic nicely) but the finer details of how all this stuff should coalesce aren't really there yet.

I'll work on an RFC for this type of service abstraction moving forward, but I think for now it's best to resist the temptation to formalize it in code just yet.

colin-grapl commented on src/rust/pipeline-ingress/src/main.rs:
```diff
@@ -0,0 +1,209 @@
+            Ok(res)
+        }
+        Err(err) => {
+            tracing::error!("configuration error: {}", err);
```
I'd suggest:
```
tracing::error!(message="configuration error", err=%err);
```
but it's not a big deal here

jgrillo-grapl replied:
fixed

colin-grapl commented on src/rust/pipeline-ingress/tests/integration_test.rs:
```diff
@@ -0,0 +1,186 @@
+
+    // we throw in a healthy sleep here to make well sure the kafka consumer is
+    // actually consuming before we throw a message at it
+    tokio::time::sleep(Duration::from_millis(5000)).await;
```
5 seconds seems really high, shouldn't it pretty much be immediately listening for messages?

jgrillo-grapl replied:
It is high locally. But I'd like to give this a good chance of succeeding in a cloud environment as well, and until we have reliable metrics I can't know what a reasonable timeout is. So I've set this (and the others) conservatively quite high with the hope that this will be a reliable test in the `testing` pipeline (e.g. AWS), and we can turn it down once we have a higher fidelity understanding of the actual environment.

colin-grapl commented on src/rust/rust-proto-new/tests/integration_test.rs:
```diff
@@ -235,18 +259,179 @@ proptest! {
+        })
+        .await
+    {
+        println!("expected error response, received: {:?}", res);
```
trace instead of print?

jgrillo-grapl replied:
fixed

jgrillo-grapl committed:
some updates for review comments

jgrillo-grapl committed:
mark lint-proto-breaking soft_fail
